### ISSUE TITLE
Deadlock in `std::thread::park`

### ISSUE THREAD

**Issue by h33p** [rust-lang/miri#3146]
I've encountered a relatively reproducible deadlock situation in my project. However, I'm not sure if it's a problem with it, because the deadlock appears to be in mpsc's recv.

- [Affected commit](https://github.com/memflow/mfio/tree/d1b28e187dbab024dfcea42d4cc5fdce65ed0ed4)
- [CI Run](https://github.com/memflow/mfio/actions/runs/6670952390/job/18131880616)

Reproduction example:

```
MIRIFLAGS="-Zmiri-seed=1 -Zmiri-backtrace=full -Zmiri-ignore-leaks -Zmiri-symbolic-alignment-check -Zmiri-retag-fields=all -Zmiri-symbolic-alignment-check -Zmiri-strict-provenance -Zmiri-disable-isolation -Zmiri-tree-borrows" cargo +nightly miri test -p mfio-rt native::tests
...
running 8 tests
test native::tests::read_all ... error: deadlock: the evaluated program deadlocked
   --> /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:35:47
    |
35  |     let r = libc::pthread_cond_wait(cond, lock);
    |                                               ^ the evaluated program deadlocked
    |
    = note: inside `std::sys::unix::thread_parking::pthread::wait` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:35:47: 35:48
    = note: inside `std::sys::unix::thread_parking::pthread::Parker::park` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:186:13: 186:51
    = note: inside `std::thread::park` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:1068:9: 1068:49
    = note: inside `std::sync::mpmc::context::Context::wait_until` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:139:17: 139:31
    = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/list.rs:444:27: 444:50
    = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:50:13: 50:18
    = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:58:31: 58:37
    = note: inside `std::thread::LocalKey::<std::cell::Cell<std::option::Option<std::sync::mpmc::context::Context>>>::try_with::<{closure@std::sync::mpmc::context::Context::with<{closure@std::sync::mpmc::list::Channel<(u64, mfio::io::BoundPacketView<mfio::io::Write>)>::recv::{closure#1}}, ()>::{closure#1}}, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:270:16: 270:31
    = note: inside `std::sync::mpmc::context::Context::with::<{closure@std::sync::mpmc::list::Channel<(u64, mfio::io::BoundPacketView<mfio::io::Write>)>::recv::{closure#1}}, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:53:9: 62:15
    = note: inside `std::sync::mpmc::list::Channel::<(u64, mfio::io::BoundPacketView<mfio::io::Write>)>::recv` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/list.rs:434:13: 455:15
    = note: inside `std::sync::mpmc::Receiver::<(u64, mfio::io::BoundPacketView<mfio::io::Write>)>::recv` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/mod.rs:307:43: 307:58
    = note: inside `std::sync::mpsc::Receiver::<(u64, mfio::io::BoundPacketView<mfio::io::Write>)>::recv` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpsc/mod.rs:849:9: 849:26
    = note: inside `<std::sync::mpsc::IntoIter<(u64, mfio::io::BoundPacketView<mfio::io::Write>)> as std::iter::Iterator>::next` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpsc/mod.rs:1075:9: 1075:23
note: inside closure
   --> mfio-rt/src/native/impls/thread.rs:254:39
    |
254 |                     for (pos, buf) in read_rx {
```

It's not a consistent deadlock, it only happens with ~10% of the seeds.

This appears to be further tweakable with `-Zmiri-preemption-rate` arg (here I also added `-Zmiri-compare-exchange-weak-failure-rate=0.0 -Zmiri-disable-weak-memory-emulation`):

```
MIRI_TEST_THREADS=1 MIRIFLAGS="-Zmiri-seed=1 -Zmiri-preemption-rate=0.001 -Zmiri-compare-exchange-weak-failure-rate=0.0 -Zmiri-disable-weak-memory-emulation -Zmiri-backtrace=full -Zmiri-ignore-leaks -Zmiri-symbolic-alignment-check -Zmiri-retag-fields=all -Zmiri-symbolic-alignment-check -Zmiri-strict-provenance -Zmiri-disable-isolation -Zmiri-tree-borrows" cargo +nightly miri test -p mfio-rt native::tests
...
running 8 tests
test native::tests::read_all ... ok
test native::tests::read_to_end ... ok
test native::tests::self_no_doublewake ... ok
test native::tests::self_wake ... ok
test native::tests::simple_io ... ok
test native::tests::wake_test_lot ... ok
test native::tests::wake_test_single ... ok
test native::tests::write_test ... error: deadlock: the evaluated program deadlocked
   --> /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:35:47
    |
35  |     let r = libc::pthread_cond_wait(cond, lock);
    |                                               ^ the evaluated program deadlocked
    |
    = note: inside `std::sys::unix::thread_parking::pthread::wait` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:35:47: 35:48
    = note: inside `std::sys::unix::thread_parking::pthread::Parker::park` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:186:13: 186:51
    = note: inside `std::thread::park` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:1068:9: 1068:49
note: inside `<std::thread::Thread as mfio::poller::ParkHandle>::park`
   --> /Users/auri/projects/mfio/mfio/src/poller.rs:94:9
    |
94  |         thread::park();
    |         ^^^^^^^^^^^^^^
```

Disabling preemption makes the problem go away. Raising preemption rate high enough makes it less reproducible. The default value seems to hit the sweet spot.

I'm not sure whether this is a true positive, because the first deadlock is basically in [channel iteration](https://github.com/memflow/mfio/blob/d1b28e187dbab024dfcea42d4cc5fdce65ed0ed4/mfio-rt/src/native/impls/thread.rs#L254), while the other is in `thread::park` call that another thread is supposed to unpark (possibly before parking is requested, but that should be okay).

Basically, is there a likely way to trigger this with bad code, without triggering other UB detection mechanisms? Also considering that no other UB was detected (with the env vars I specified).

In addition, how would one go about deconstructing this into a minimal sample?

**Comment by RalfJung** [comment]
> no other UB

Note that deadlocks aren't UB! If you get a deadlock error then no UB was detected in the execution until then (and the execution stops here so no UB can occur later either).

> In addition, how would one go about deconstructing this into a minimal sample?

Good question. The more you can reduce the amount of code needed to trigger the issue (including using less of the standard library), the better. But there's no general recipe.

We could also potentially make the deadlock error more useful, if we can figure out which information would be good to print. Right now we are showing the backtrace of the last thread that took a step and went to sleep. It seems to me it might be good to print the current span of *all* threads, so that one can try to get an idea how one would arrive in a situation where they are all sleeping. I'm a bit worried someone has 500 threads and this error becomes unreadable, so maybe we should truncate at ~10 threads. Do you think that would be useful?

**Comment by bjorn3** [comment]
It would be nice if miri could know for every thread blocking on a mutex acquire (or other sync primitive) which thread had acquired the mutex. That would allow it to walk the graph of which thread waits on which thread and find the cycle to show the user as well as directly show the thread a thread waits on to the user as debugging aid.

**Comment by RalfJung** [comment]
Yeah that's the fancy version, that could even detect deadlocks when some threads are still running but there is a cycle of mutex dependencies. That's going to be a bunch of work though. "Show where all threads are sleeping" is something I can probably do fairly quickly.

**Comment by h33p** [comment]
> > no other UB
> 
> Note that deadlocks aren't UB! If you get a deadlock error then no UB was detected in the execution until then (and the execution stops here so no UB can occur later either).
> 
> > In addition, how would one go about deconstructing this into a minimal sample?
> 
> Good question. The more you can reduce the amount of code needed to trigger the issue (including using less of the standard library), the better. But there's no general recipe.

Thanks a lot for a quick response!

> We could also potentially make the deadlock error more useful, if we can figure out which information would be good to print. Right now we are showing the backtrace of the last thread that took a step and went to sleep. It seems to me it might be good to print the current span of _all_ threads, so that one can try to get an idea how one would arrive in a situation where they are all sleeping. I'm a bit worried someone has 500 threads and this error becomes unreadable, so maybe we should truncate at ~10 threads. Do you think that would be useful?

That would probably be really useful, I managed to create a relatively consistent repro that is doing some obviously flawed things (see details).

<details>

```rust
use std::sync::mpsc::{self, Receiver, Sender};
use std::thread::*;

fn sleep_ms(millis: usize) {
    for _ in 0..(millis * 10) {
        core::hint::spin_loop();
    }
}

fn echo_thread(rx1: Receiver<(usize, Thread)>, tx2: Sender<usize>) {
    for (v, t) in rx1 {
        t.unpark();
        if tx2.send(v).is_err() {
            break;
        }
        sleep_ms(10);
    }
}

fn test(elems: usize) {
    let (tx1, rx1) = mpsc::channel();
    let (tx2, rx2) = mpsc::channel();

    let jh = spawn(move || echo_thread(rx1, tx2));

    let t = current();

    for i in 0..elems {
        tx1.send((i, t.clone())).unwrap();
    }
        
    park();
    park();
    park();
    park();

    for i in 0..elems {
        let v = rx2.recv().unwrap();
        assert_eq!(i, v);
    }

    core::mem::drop(tx1);
    core::mem::drop(rx2);

    jh.join().unwrap();
}

fn main() {
    let threads = (0..2).map(|_| spawn(move || test(10))).collect::<Vec<_>>();
    for t in threads {
        t.join().unwrap();
    }
}
```

The idea is processing thread unparks the caller thread on every iteration, and each iteration takes a bit to get processed, potentially allowing for the 4 thread parks to pass without causing deadlock. However, we have the following case:

```
❯ time MIRIFLAGS="-Zmiri-seed=54 -Zmiri-backtrace=full  -Zmiri-ignore-leaks -Zmiri-symbolic-alignment-check -Zmiri-retag-fields=all -Zmiri-symbolic-alignment-check -Zmiri-strict-provenance -Zmiri-disable-isolation -Zmiri-tree-borrows" cargo +nightly miri run
Preparing a sysroot for Miri (target: aarch64-apple-darwin)... done
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `/Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/bin/cargo-miri runner target/miri/aarch64-apple-darwin/debug/parklock`
error: deadlock: the evaluated program deadlocked
  --> /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:35:47
   |
35 |     let r = libc::pthread_cond_wait(cond, lock);
   |                                               ^ the evaluated program deadlocked
   |
   = note: inside `std::sys::unix::thread_parking::pthread::wait` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:35:47: 35:48
   = note: inside `std::sys::unix::thread_parking::pthread::Parker::park` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread_parking/pthread.rs:186:13: 186:51
   = note: inside `std::thread::park` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:1068:9: 1068:49
   = note: inside `std::sync::mpmc::context::Context::wait_until` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:139:17: 139:31
   = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/list.rs:444:27: 444:50
   = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:50:13: 50:18
   = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:58:31: 58:37
   = note: inside `std::thread::LocalKey::<std::cell::Cell<std::option::Option<std::sync::mpmc::context::Context>>>::try_with::<{closure@std::sync::mpmc::context::Context::with<{closure@std::sync::mpmc::list::Channel<(usize, std::thread::Thread)>::recv::{closure#1}}, ()>::{closure#1}}, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/local.rs:270:16: 270:31
   = note: inside `std::sync::mpmc::context::Context::with::<{closure@std::sync::mpmc::list::Channel<(usize, std::thread::Thread)>::recv::{closure#1}}, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/context.rs:53:9: 62:15
   = note: inside `std::sync::mpmc::list::Channel::<(usize, std::thread::Thread)>::recv` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/list.rs:434:13: 455:15
   = note: inside `std::sync::mpmc::Receiver::<(usize, std::thread::Thread)>::recv` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpmc/mod.rs:307:43: 307:58
   = note: inside `std::sync::mpsc::Receiver::<(usize, std::thread::Thread)>::recv` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpsc/mod.rs:849:9: 849:26
   = note: inside `<std::sync::mpsc::IntoIter<(usize, std::thread::Thread)> as std::iter::Iterator>::next` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sync/mpsc/mod.rs:1075:9: 1075:23
note: inside `echo_thread`
  --> src/main.rs:11:19
   |
11 |     for (v, t) in rx1 {
   |                   ^^^
note: inside closure
  --> src/main.rs:24:28
   |
24 |     let jh = spawn(move || echo_thread(rx1, tx2));
   |                            ^^^^^^^^^^^^^^^^^^^^^
   = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<{closure@src/main.rs:24:20: 24:27}, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:154:18: 154:21
   = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:529:17: 529:78
   = note: inside `<std::panic::AssertUnwindSafe<{closure@std::thread::Builder::spawn_unchecked_<'_, '_, {closure@src/main.rs:24:20: 24:27}, ()>::{closure#1}::{closure#0}}> as std::ops::FnOnce<()>>::call_once` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9: 271:19
   = note: inside `std::panicking::r#try::do_call::<std::panic::AssertUnwindSafe<{closure@std::thread::Builder::spawn_unchecked_<'_, '_, {closure@src/main.rs:24:20: 24:27}, ()>::{closure#1}::{closure#0}}>, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:504:40: 504:43
   = note: inside `std::panicking::r#try::<(), std::panic::AssertUnwindSafe<{closure@std::thread::Builder::spawn_unchecked_<'_, '_, {closure@src/main.rs:24:20: 24:27}, ()>::{closure#1}::{closure#0}}>>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/panicking.rs:468:19: 468:81
   = note: inside `std::panic::catch_unwind::<std::panic::AssertUnwindSafe<{closure@std::thread::Builder::spawn_unchecked_<'_, '_, {closure@src/main.rs:24:20: 24:27}, ()>::{closure#1}::{closure#0}}>, ()>` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/panic.rs:142:14: 142:33
   = note: inside closure at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/thread/mod.rs:528:30: 530:16
   = note: inside `<{closure@std::thread::Builder::spawn_unchecked_<'_, '_, {closure@src/main.rs:24:20: 24:27}, ()>::{closure#1}} as std::ops::FnOnce<()>>::call_once - shim(vtable)` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5: 250:71
   = note: inside `<std::boxed::Box<dyn std::ops::FnOnce()> as std::ops::FnOnce<()>>::call_once` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2007:9: 2007:52
   = note: inside `<std::boxed::Box<std::boxed::Box<dyn std::ops::FnOnce()>> as std::ops::FnOnce<()>>::call_once` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2007:9: 2007:52
   = note: inside `std::sys::unix::thread::Thread::new::thread_start` at /Users/auri/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17: 108:64

error: aborting due to previous error
```
</details>

What you may see in details is the similar deadlock on mpsc iteration, which is probably not the most useful piece of information - the source of the deadlock is one of the spurious `thread::park()` calls. Seeing both would be absolutely great.

I suppose the first step would be to try to look into multiple thread spans, and print them. I'll try to figure that one out. As you said, would not be the most pleasant flag to use with many threads existing, but would likely be useful to debug these complex cases if some level of minimization is possible.

**Comment by RalfJung** [comment]
> doing some obviously flawed things (see details).

So is a deadlock expected because the code is flawed? Or what do you mean?

**Comment by RalfJung** [comment]
> I suppose the first step would be to try to look into multiple thread spans, and print them. I'll try to figure that one out.

Great. :) Let us know if you have any questions. (Might be good to ask those on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/269128-miri) to avoid flooding this issue with those questions.)

**Comment by h33p** [comment]
> > doing some obviously flawed things (see details).
> 
> So is a deadlock expected because the code is flawed? Or what do you mean?

Yup, exactly.

**Comment by RalfJung** [comment]
Yeah parking multiple times in a row is definitely a potential cause of deadlocks. "unpark" tickets do no accumulate. So it is crucial that *between one park() and the next*, you arrange for another thread to call "unpark".

**Comment by RalfJung** [comment]
I'll close this as I don't think there is a Miri bug here. But we could certainly do better in our deadlock diagnostics, which is now tracked at https://github.com/rust-lang/miri/issues/3424.
